### DISCUSSION TITLE
Why is the HttpClient constructor not internal?

### DISCUSSION THREAD

**Discussion by CraigComeOnThisNameCantBeTaken** [dotnet/runtime#59343] [Ideas] [answered]
I often see it used as a concrete instantiation with a using statement, but this doesn't manage sockets and leads to stale DNS lookups and socket exhaustion.

What is the reason for allowing this usage please? If it were internal developers could not make that mistake.

**Comment by huoyaoyuan** [comment, marked as answer]
> but this doesn't manage sockets and leads to stale DNS lookups and socket exhaustion.

Socket is managed by connection pool which is used by `HttpClientHandler`. Each `HttpClient` has its own connection pool.

Although it's typically shared during application lifetime, short-lived ones are also necessary for application which only uses network very rarely.

**Reply by CraigComeOnThisNameCantBeTaken** [reply]
Thanks

**Reply by CraigComeOnThisNameCantBeTaken** [reply]
Am I correct that the HttpClientHandler doesnt release sockets when HttpClient is disposed?

**Reply by huoyaoyuan** [reply]
`HttpClient.Dispose` calls `HttpClientHandler.Dispose`, which finally calls `HttpConnectionHandler.Dispose`, which calls `HttpCollectionPoolManager.Dispose`, which disposes each `HttpConnectionPool`, which disposes `HttpConnection`, which disposes sockets eventually.

The sockets are wrapped into `NetworkStream`, so you need to search for `Stream.Dispose` in `System.Net.Http` project. The disposal for non-exceptional path should be here:

https://github.com/dotnet/runtime/blob/8c553ad3542b52909ff42d0cd278c766f1112e17/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L133
